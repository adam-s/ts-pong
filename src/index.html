<!DOCTYPE html> 
<html>
<head>
  <title>Pong!</title> 
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/> 
  <style href="pong.css" media="screen, print" rel="stylesheet" type="text/css"></style> 
</head> 
 
<body> 

  <div id="sidebar">

    <h2>This is Pong!</h2>

    <div class='description'>
      <p>
        This is a javascript version of Pong.
      </p>
      <p>
        Press <b>1</b> for a single player game.<br>
        Press <b>2</b> for a double player game.<br>
        Press <b>0</b> to watch the computer play itself.
      </p>
      <p>
        Player 1 moves using the <b>Q</b> and <b>A</b> keys.<br>
        Player 2 moves using the <b>P</b> and <b>L</b> keys.
      </p>
      <p>
        <b>Esc</b> can be used to abandon a game.
      </p>
    </div>

    <div class='settings'>
      <label for='sound'>sound: </label>
      <input type='checkbox' id='sound'>
    </div>

    <div class='settings'>
      <label for='stats'>stats: </label>
      <input type='checkbox' id='stats' checked>
    </div>

    <div class='settings'>
      <label for='footprints'>footprints: </label>
      <input type='checkbox' id='footprints'>
    </div>

    <div class='settings'>
      <label for='predictions'>predictions: </label>
      <input type='checkbox' id='predictions'>
    </div>

  </div>

  <canvas id="game">
    <div id="unsupported">
      Sorry, this example cannot be run because your browser does not support the &lt;canvas&gt; element
    </div> 
  </canvas>
</body> 
</html>
