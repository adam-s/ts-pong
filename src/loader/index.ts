export * from './Loader';
export * from './Resource';
export * from './ImageSource';
export * from './AudioSource';
export * from './AudioContext';
export * from './WebAudioInstance';
