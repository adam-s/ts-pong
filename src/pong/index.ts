export * from './Pong';
export * from './Ball';
export * from './Court';
export * from './Menu';
export * from './Paddle';
export * from './resources';
