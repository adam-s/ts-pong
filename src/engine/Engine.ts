/* eslint-disable @typescript-eslint/no-empty-function */
import { Loadable, Loader } from '../loader';
import { timestamp, PartialBy } from '../utils';

export interface Game {
  init?: () => void;
  width: number;
  height: number;
  start: (numPlayers: number) => void;
  stop: () => void;
  update: (dt: number) => void;
  draw: (ctx: CanvasRenderingContext2D) => void;
  onkeyup?: (ev: KeyboardEvent) => void;
  onkeydown?: (ev: KeyboardEvent) => void;
}

export interface IGameOptions {
  fps?: number;
  stats?: boolean;
  width?: number;
  height?: number;
  canvasElementId?: string;
}

const defaultOptions: IGameOptions = {
  fps: 60,
  stats: true,
  width: 640,
  height: 480,
  canvasElementId: 'game',
};

export class Engine {
  private _options: IGameOptions;

  private _timer: ReturnType<typeof setInterval>;

  private _lastFrame: number;

  private _fps: number;

  private _interval: number;

  private _canvasElementId: string;

  private _width: number;

  private _height: number;

  private _front: HTMLCanvasElement;

  private _back: HTMLCanvasElement;

  private _front2d: CanvasRenderingContext2D;

  private _back2d: CanvasRenderingContext2D;

  private _loader: Loader;

  private _isLoading = false;

  private _hasStarted = false;

  private _stats = {
    count: 0,
    fps: 0,
    update: 0,
    draw: 0,
    frame: 0, // update + draw
  };

  constructor(
    private game: Game,
    options?: PartialBy<IGameOptions, 'canvasElementId'>
  ) {
    // For some inspiration how to include options to a TypeScript class
    // https://github.com/microsoft/vscode/blob/main/src/vs/editor/contrib/zoneWidget/zoneWidget.ts
    this._options = { ...defaultOptions, ...options };

    this._fps = this._options.fps;
    this._interval = 1000.0 / this._options.fps;

    // Set or create the front canvas element
    if (this._options.canvasElementId) {
      this._front = <HTMLCanvasElement>(
        document.getElementById(this._options.canvasElementId)
      );
    } else {
      this._front = <HTMLCanvasElement>document.createElement('canvas');
    }

    // Create a back canvas element
    this._back = <HTMLCanvasElement>document.createElement('canvas');

    this._width = this.game.width;
    this._height = this.game.height;

    this._front.width = this._back.width = this._width;
    this._front.height = this._back.height = this._height;

    this._front2d = this._front.getContext('2d');
    this._back2d = this._back.getContext('2d');

    this._loader = new Loader();

    this.addEvents();
  }

  async start(loader?: Loader): Promise<void> {
    // if (!this._compatible) {
    //   return Promise.reject('This game is incompatible with your browser');
    // }
    let loadingComplete: Promise<void>;
    if (loader) {
      this._loader = loader;
      loadingComplete = this.load(this._loader);
    } else {
      loadingComplete = Promise.resolve();
    }
    await loadingComplete;

    // Give an options to call and async init in the case the game requires loading resources
    this.game.init?.();

    this._lastFrame = timestamp();
    this._timer = setInterval(() => {
      this.loop();
    }, this._interval);
  }

  public load(loader: Loadable<unknown>): Promise<void> {
    const complete = new Promise<void>((resolve) => {
      this._isLoading = true;
      loader.load().then(() => {
        this._isLoading = false;
        resolve();
      });
    });

    return complete;
  }

  stop(): void {
    clearInterval(this._timer);
  }

  loop(): void {
    const start = timestamp();
    this.update((start - this._lastFrame) / 1000.0); // send dt as seconds

    const middle = timestamp();
    this.draw();

    const end = timestamp();
    this.updateStats(middle - start, end - middle);
    this._lastFrame = start;
  }

  update(dt: number): void {
    // might as well let the engine watch key press and inform the game on every update rather than pass through
    // with an eventing
    this.game.update(dt);
  }

  draw(): void {
    this._back2d.clearRect(0, 0, this._width, this._height);
    this.game.draw(this._back2d);
    this.drawStats(this._back2d);
    this._front2d.clearRect(0, 0, this._width, this._height);
    this._front2d.drawImage(this._back, 0, 0);
  }

  resetStats() {
    this._stats = {
      count: 0,
      fps: 0,
      update: 0,
      draw: 0,
      frame: 0, // update + draw
    };
  }

  updateStats(update: number, draw: number) {
    if (this._options.stats) {
      this._stats.update = Math.max(1, update);
      this._stats.draw = Math.max(1, draw);
      this._stats.frame = this._stats.update + this._stats.draw;
      this._stats.count =
        this._stats.count == this._fps ? 0 : this._stats.count + 1;
      this._stats.fps = Math.min(this._fps, 1000 / this._stats.frame);
    }
  }

  drawStats(ctx: CanvasRenderingContext2D) {
    if (this._options.stats) {
      ctx.fillText(
        'frame: ' + this._stats.count,
        this._width - 100,
        this._height - 60
      );
      ctx.fillText(
        'fps: ' + this._stats.fps,
        this._width - 100,
        this._height - 50
      );
      ctx.fillText(
        'update: ' + this._stats.update + 'ms',
        this._width - 100,
        this._height - 40
      );
      ctx.fillText(
        'draw: ' + this._stats.draw + 'ms',
        this._width - 100,
        this._height - 30
      );
    }
  }

  addEvents() {
    document.addEventListener('keyup', this.onkeyup.bind(this), false);
    document.addEventListener('keydown', this.onkeydown.bind(this), false);
  }

  onkeyup(ev: KeyboardEvent) {
    if (this.game.onkeyup) this.game.onkeyup(ev);
  }

  onkeydown(ev: KeyboardEvent) {
    if (this.game.onkeydown) this.game.onkeydown(ev);
  }

  hideCurser() {
    this._front.style.cursor = 'none';
  }

  showCurser() {
    this._front.style.cursor = 'auto';
  }

  alert(msg: string) {
    this.stop(); // alert blocks thread, so need to stop game loop in order to avoid sending huge dt values to next update
    const result = window.alert(msg);
    this.start();
    return result;
  }

  confirm(msg: string) {
    this.stop(); // alert blocks thread, so need to stop game loop in order to avoid sending huge dt values to next update
    const result = window.confirm(msg);
    this.start();
    return result;
  }

  showStats(show: boolean) {
    this._options.stats = show;
  }

  static random(min: number, max: number) {
    return min + Math.random() * (max - min);
  }
}
