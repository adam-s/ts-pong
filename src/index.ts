import './styles.scss';
import { Engine } from './engine/Engine';
import { Pong, loader } from './pong';

const pong = new Pong();
const engine = new Engine(pong, { canvasElementId: 'game' });

engine.start(loader).then(() => {
  const stats = document.getElementById('stats') as HTMLInputElement;
  const sound = document.getElementById('sound') as HTMLInputElement;
  const footprints = document.getElementById('footprints') as HTMLInputElement;
  const predictions = document.getElementById(
    'predictions'
  ) as HTMLInputElement;

  // engine.showStats(stats.checked);
  // pong.playSounds(sound.checked);
  // pong.showFootprints(footprints.checked);
  // pong.showPredictions(predictions.checked);

  stats.addEventListener('change', () => engine.showStats(stats.checked));
  sound.addEventListener('change', () => pong.playSounds(sound.checked));
  footprints.addEventListener('change', () =>
    pong.showFootprints(footprints.checked)
  );
  predictions.addEventListener('change', () =>
    pong.showPredictions(predictions.checked)
  );
});
