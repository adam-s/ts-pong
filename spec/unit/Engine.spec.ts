import 'jest-canvas-mock';
import { Engine, Game, IGameOptions } from '../../src/engine';

describe('Engine', () => {
  it('exists', () => {
    expect(Engine).toBeDefined();
  });

  const gameMock: Game = {
    width: 0,
    height: 0,
    start: () => null,
    stop: () => null,
    update: () => null,
    draw: () => null,
  };

  const options: IGameOptions = {
    fps: 0,
    stats: true,
    width: 600,
    height: 400,
  };

  it('has default values', () => {
    const implementation = new Engine(gameMock);
    expect(implementation['_options'].fps).toEqual(60);
  });

  it('overrides default values', () => {
    const implementation = new Engine(gameMock, options);
    expect(implementation['_options'].fps).toEqual(0);
  });

  it('creates a front and back canvas element', () => {
    const implementation = new Engine(gameMock, options);
    expect(implementation['_front']).toBeDefined();
    expect(implementation['_back']).toBeDefined();
  });

  it('creates a canvas element if a canvasElementId is passed in options', () => {
    // Initialize the DOM
    const canvasElementId = 'test';
    document.body.innerHTML = `<div><canvas id="${canvasElementId}"></canvas></div>`;
    const optionsWithCanvasElementId = Object.assign({}, options, {
      canvasElementId,
    });
    const implementation = new Engine(gameMock, optionsWithCanvasElementId);
    expect(implementation['_front']).toBeDefined();
  });

  it('renders stats to canvas', () => {
    const canvas = document.createElement('canvas');
    const ctx = canvas.getContext('2d');

    const implementation = new Engine(gameMock, options);

    implementation.drawStats(ctx);
    const events = ctx.__getEvents();
    expect(events).toMatchSnapshot();
    ctx.__clearEvents();
  });

  it('adds events', () => {
    const documentSpy = jest
      .spyOn(document, 'addEventListener')
      .mockImplementation(() => undefined);
    new Engine(gameMock, options);
    expect(documentSpy.mock.calls[0][0]).toEqual('keyup');
    expect(documentSpy.mock.calls[1][0]).toEqual('keydown');
  });
});
