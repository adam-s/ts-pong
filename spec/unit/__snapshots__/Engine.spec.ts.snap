// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engine renders stats to canvas 1`] = `
Array [
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "frame: 0",
      "x": -100,
      "y": -60,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "fps: 0",
      "x": -100,
      "y": -50,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "update: 0ms",
      "x": -100,
      "y": -40,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "draw: 0ms",
      "x": -100,
      "y": -30,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;
